node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/NodeAlerter.java

```diff
@@ -0,0 +1,442 @@
+        return overcommittedNodes;
+    }
+
+    private Map<Node, List<Allocation>> collateAllocations(Map<Node, List<Node>> nodeChildren) {
+        return nodeChildren.entrySet().stream().collect(Collectors.toMap(
+                Map.Entry::getKey,
+                e -> e.getValue().stream()
+                        .map(Node::allocation).flatMap(Optional::stream)
+                        .collect(Collectors.toList())
+        ));
+    }
+
+    /**
+     * Data class used for detailing why removing the given tenant from the given host was unsuccessful.
+     * A failure might not be caused by failing to allocate a specific tenant, in which case the fields
+     * will be empty.
+     */
+    public static class HostRemovalFailure {
+        Optional<Node> host;
+        Optional<Node> tenant;
+        AllocationFailureReasonList failureReasons;
+        public static HostRemovalFailure none() {
+            return new HostRemovalFailure(
+                    Optional.empty(),
+                    Optional.empty(),
+                    new AllocationFailureReasonList(List.of()));
+        }
+        public static HostRemovalFailure create(Node host, Node tenant, AllocationFailureReasonList failureReasons) {
+            return new HostRemovalFailure(
+                    Optional.of(host),
+                    Optional.of(tenant),
+                    failureReasons);
+        }
+        private HostRemovalFailure(Optional<Node> host, Optional<Node> tenant, AllocationFailureReasonList failureReasons) {
+            this.host = host;
+            this.tenant = tenant;
+            this.failureReasons = failureReasons;
+        }
+    }
+
+    /**
+     * Tests whether it's possible to remove the provided hosts.
+     * Does not mutate any input variable.
+     * @return Empty optional if removal is possible, information on what caused the failure otherwise
+     */
+    private Optional<HostRemovalFailure> findHostRemovalFailure(List<Node> hostsToRemove, List<Node> allHosts,
+                                                        Map<Node, List<Node>> nodechildren,
+                                                        Map<Node, AllocationResources> availableResources) {
+        var containedAllocations = collateAllocations(nodechildren);
+        var resourceMap = new HashMap<>(availableResources);
+        List<Node> validAllocationTargets = allHosts.stream()
+                .filter(h -> !hostsToRemove.contains(h))
+                .collect(Collectors.toList());
+        if (validAllocationTargets.size() == 0) {
+            return Optional.of(HostRemovalFailure.none());
+        }
+
+        for (var host : hostsToRemove) {
+            Optional<Node> unallocatedNode = tryAllocateNodes(nodechildren.get(host),
+                    validAllocationTargets, resourceMap, containedAllocations);
+
+            if (unallocatedNode.isPresent()) {
+                return Optional.of(HostRemovalFailure.create(host, unallocatedNode.get(),
```

Comment:
That's a lot of function calls within function calls. Might be more readable if you create and pass around an `AllocationFailureReasonList`object.